File: c/leet_code/678.c
#include <stdbool.h>

bool checkValidString(const char *s) {
    int low = 0, high = 0;
    for (int i = 0; s[i]; i++) {
        if (s[i] == '(') {
            low++;
            high++;
        } else if (s[i] == ')') {
            low--;
            high--;
        } else {
            low--;
            high++;
        }
        if (high < 0) {
            return false;
        }
        low = low > 0 ? low : 0;
    }
    return low == 0;
}

File: php/leet_code/678.php
<?php

class Solution
{

    /**
     * @param String $s
     * @return Boolean
     */
    function checkValidString(string $s): bool
    {
        $low = 0;
        $high = 0;
        for ($i = 0, $iMax = strlen($s); $i < $iMax; $i++) {
            if ($s[$i] === '(') {
                $low++;
                $high++;
            } else if ($s[$i] === ')') {
                $low--;
                $high--;
            } else {
                $low--;
                $high++;
            }
            if ($high < 0) {
                return false;
            }
            $low = max($low, 0);
        }
        return $low === 0;
    }
}
